libraries/apollo-websocket-network-transport-incubating/src/jsMain/kotlin/com/apollographql/apollo3/network/websocket/WebSocketEngine.js.kt

```diff
@@ -0,0 +1,170 @@
+            platformWebSocket.close(CLOSE_GOING_AWAY.toShort(), "Unsupported message received")
+          }
+          Unit
+        }
+      }
+    }
+
+    platformWebSocket.onerror = {
+      if (!disposed) {
+        disposed = true
+        listener.onError(DefaultApolloException("Error while reading websocket"))
+      }
+    }
+
+    platformWebSocket.onclose = {
+      val event: dynamic = it
+      if (!disposed) {
+        disposed = true
+        if (event.wasClean) {
+          listener.onClosed(event.code, event.reason)
+        } else {
+          listener.onError(DefaultApolloException("WebSocket was closed"))
+        }
+      }
+    }
+  }
+
+  override fun send(data: ByteArray) {
+    check(::platformWebSocket.isInitialized) {
+      "You must call connect() before send()"
+    }
+
+    if (platformWebSocket.bufferedAmount.toDouble() > MAX_BUFFERED) {
+      if (!disposed) {
+        disposed = true
+        listener.onError(DefaultApolloException("Too much data queued"))
+      }
+    }
+    platformWebSocket.send(Uint8Array(data.toTypedArray()))
+  }
+
+  override fun send(text: String) {
+    check(::platformWebSocket.isInitialized) {
+      "You must call connect() before send()"
+    }
+
+    if (platformWebSocket.bufferedAmount.toDouble() > MAX_BUFFERED) {
+      if (!disposed) {
+        disposed = true
+        listener.onError(DefaultApolloException("Too much data queued"))
+      }
+    }
+
+    platformWebSocket.send(text)
+  }
+
+  override fun close(code: Int, reason: String) {
+    check(::platformWebSocket.isInitialized) {
+      "You must call connect() before close()"
+    }
+
+    if (!disposed) {
+      disposed = true
+      platformWebSocket.close(code.toShort(), reason)
+    }
+  }
+}
+
+/**
+ * This is probably far too high and problems will probably appear much sooner but this is used to signal the intent
+ * of this code as well as a last resort
+ */
+private val MAX_BUFFERED = 100_000_000
+
+/**
+ * Copyright 2014-2019 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ * From https://github.com/ktorio/ktor/blob/6cd529b2dcedfcfc4ca2af0f62704764e160d7fd/ktor-utils/js/src/io/ktor/util/PlatformUtilsJs.kt#L16
+ */
+fun isNode(): Boolean {
+  return js(
+      """
+                (typeof process !== 'undefined' 
+                    && process.versions != null 
+                    && process.versions.node != null) ||
+                (typeof window !== 'undefined' 
+                    && typeof window.process !== 'undefined' 
+                    && window.process.versions != null 
+                    && window.process.versions.node != null)
+                """
+  ) as Boolean
+}
+
+/*
+ * The function below works due to ktor being used for the HTTP engine
+ * and how ktor imports ws, if this changes, ws would need to be added
+ * as a direct dependency.
```

Comment:
I think this comment can just be removed?

Reply:
100%